internal/core/input.rs

```diff
@@ -486,105 +486,98 @@ pub fn process_mouse_input(
     }
 
     let mut result = MouseInputState::default();
-    type State = (Vector2D<Coord>, Vec<(ItemWeak, InputEventFilterResult)>, MouseEvent);
-    crate::item_tree::visit_items_with_post_visit(
-        &component,
-        crate::item_tree::TraversalOrder::FrontToBack,
-        |comp_rc: &ComponentRc,
-         item: core::pin::Pin<ItemRef>,
-         item_index: usize,
-         (offset, mouse_grabber_stack, mouse_event): &State| {
-            let item_rc = ItemRc::new(comp_rc.clone(), item_index);
-
-            let mut mouse_event = *mouse_event;
-
-            let geom = item.as_ref().geometry();
-            let geom = geom.translate(*offset);
-
-            let mut mouse_grabber_stack = mouse_grabber_stack.clone();
-
-            let post_visit_state = if mouse_event.position().map_or(false, |p| geom.contains(p))
-                || crate::item_rendering::is_clipping_item(item)
-            {
-                let mut event2 = mouse_event;
-                event2.translate(-geom.origin.to_vector());
-                let filter_result = item.as_ref().input_event_filter_before_children(
+    let root = ItemRc::new(component.clone(), 0);
+    send_mouse_event_to_item(mouse_event, root, window_adapter, &mut result);
+
+    send_exit_events(&mouse_input_state, mouse_event.position(), window_adapter);
+
+    result
+}
+
+fn send_mouse_event_to_item(
+    mouse_event: MouseEvent,
+    item_rc: ItemRc,
+    window_adapter: &Rc<dyn WindowAdapter>,
+    result: &mut MouseInputState,
+) -> VisitChildrenResult {
+    let item = item_rc.borrow();
+    let geom = item.as_ref().geometry();
+    let mut event2 = mouse_event;
+    event2.translate(-geom.origin.to_vector());
+
+    let filter_result = if mouse_event.position().map_or(false, |p| geom.contains(p))
+        || crate::item_rendering::is_clipping_item(item)
+    {
+        item.as_ref().input_event_filter_before_children(event2, window_adapter, &item_rc)
+    } else {
+        InputEventFilterResult::ForwardAndIgnore
+    };
+
+    let (forward_to_children, ignore) = match filter_result {
+        InputEventFilterResult::ForwardEvent => (true, false),
+        InputEventFilterResult::ForwardAndIgnore => (true, true),
+        InputEventFilterResult::ForwardAndInterceptGrab => (true, false),
+        InputEventFilterResult::Intercept => (false, false),
+        InputEventFilterResult::InterceptAndDispatch(new_event) => {
+            event2 = new_event;
```

Comment:
I had to read over this at least ten times, seems correct to me. Super tricky in the old version, better here. Suggestion: `event_for_children` instead of `event2`

Reply:
i'll rename them in a future commit